### ISSUE TITLE
[Proposal] mat:eigen:power method

### ISSUE THREAD

**Issue by Konstantin8105** [gonum/gonum#591]
### What are you trying to do?
Calculate only a few first largest eigenvalues(not all).

### What did you do?
Use `gonum.org/v1/gonum/mat/eigen.go`.

### What did you expect to happen?
For example, implementation of `power method` and `deflation method`.
1) Have you a plans for implementation that methods?
2) Is acceptable to create PR with implementation?

Some links:
* `wiki` https://en.wikipedia.org/wiki/Power_iteration
* `example power method` https://nptel.ac.in/courses/111107062/module3/lecture1/lecture1.pdf http://mlwiki.org/index.php/Power_Iteration

### What actually happened?
That is not yet implemented.

### What version of Go and Gonum are you using?
`gonum.org/v1/gonum`

### Does this issue reproduce with the current master?
??

**Comment by vladimir-ch** [comment]
You are right, we only have algorithms that compute all eigenvalues of symmetric and general matrices.  I'm skeptical about implementing textbook algorithms, especially the Power method. Any practical implementation is always much more complicated. For example, for general matrices we have the implicit QR iteration which is a more sophisticated variation on the power method, and translating the corresponding LAPACK routines from Fortran to Go was a half-year effort. Do you already have any plans or proposals for an implementation? Are you interested in symmetric or general matrices? Non-symmetric matrices have in general complex eigenvalues and the concept of "largest" is not clearly defined for them.

**Comment by Konstantin8105** [comment]
In general, present state of project about calculation Ax=b, Ax=lBx is enough for my project. All your questions is very important and very huge for implementation on Go, so in my point of view, the best way is use exist software.
In my plan for future, creating a golang API of exist on C software [`lis`](https://github.com/anishida/lis) with many implementation of sparse solvers.
If it interesting for you, then just do not close that issue.

**Comment by btracey** [comment]
I'm confused. Do you mean Ax = \lambda x? Ax = b is just a solve. Also, your lapack link above is about full eigenvalue problems, and not  just largest, or am I missing something? I ask because if there is already a Lapack implementation then adding our own might not be that difficult (still might be though). 

If you have a large-sparse system then using alternate approaches is best (unless you'd like to develop them and we could potentially help).

**Comment by vladimir-ch** [comment]
In LAPACK there is
```
DSTEBZ computes the eigenvalues of a symmetric tridiagonal matrix T.
The user may ask for all eigenvalues, all eigenvalues in the half-open interval
(VL, VU], or the IL-th through IU-th eigenvalues.
```
This could be used for symmetric matrices which would have to be first reduced to the tridiagonal form. I'm not sure whether it's worth it, at least not without understanding the requirements and confirmation that getting all eigenvalues is too slow for @Konstantin8105 application (how large are the matrices?).

If the matrices are large (and possibly sparse), then an approach based on Arnoldi iteration could be taken. There is ARPACK that does that but I doubt that we have the free time to port that.

> ... creating a golang API of exist on C software lis with many implementation of sparse solvers.

There is our `exp/linsolve` package where I'm slowly working on iterative solvers in pure Go (no need to use cgo).
